bench/bench.ml

```diff
@@ -0,0 +1,484 @@
+    "Configuration:@\n    @[%a@]@\n@\nResults:@\n    @[%a@]@\n" pp_config config
+    Fmt.(list ~sep:(any "@\n@\n") pp_bench)
+    results
+
+let print_json fmt (config, results) =
+  let open Yojson.Safe in
+  let obj =
+    `Assoc
+      [
+        ("config", config_to_yojson config);
+        ( "results",
+          `Assoc
+            (List.map
+               (fun (b, result) ->
+                 (b.Index.name, Benchmark.result_to_yojson result))
+               results) );
+      ]
+  in
+  pretty_print fmt obj
+
+let run filter root seed with_metrics log_size nb_entries json =
+  let config =
+    { key_size; value_size; nb_entries; log_size; seed; with_metrics }
+  in
+  cleanup root;
+  init config;
+  let name_filter =
+    match filter with None -> fun _ -> true | Some re -> Re.execp re
+  in
+  Index.suite
+  |> schedule name_filter
+  |> List.map (fun (b : Index.suite_elt) ->
+         let name =
+           match b.dependency with None -> b.name | Some name -> name
+         in
+         let result =
+           Index.run ~with_metrics ~nb_entries ~log_size ~root ~name
+             ~fresh:b.fresh ~readonly:b.readonly b.benchmark
+         in
+         (b, result))
+  |> fun results ->
+  Fmt.pr "%a" (if json then print_json else print) (config, results)
+
+open Cmdliner
+
+let env_var s = Arg.env_var ("INDEX_BENCH_" ^ s)
+
+let regex =
+  let parse s =
+    try Ok Re.(compile @@ Pcre.re s) with
+    | Re.Perl.Parse_error -> Error (`Msg "Perl-compatible regexp parse error")
+    | Re.Perl.Not_supported -> Error (`Msg "unsupported regexp feature")
+  in
+  let print = Re.pp_re in
+  Arg.conv (parse, print)
+
+let name_filter =
+  let doc = "A regular expression matching the names of benchmarks to run" in
+  let env = env_var "NAME_FILTER" in
+  Arg.(
+    value
+    & opt (some regex) None
+    & info [ "f"; "filter" ] ~env ~doc ~docv:"NAME_REGEX")
+
+let data_dir =
+  let doc = "Set directory for the data files" in
+  let env = env_var "DATA_DIR" in
+  Arg.(value & opt dir "_bench" & info [ "d"; "data_dir" ] ~env ~doc)
+
+let seed =
+  let doc = "The seed used to generate random data." in
+  let env = env_var "SEED" in
+  Arg.(value & opt int 0 & info [ "s"; "seed" ] ~env ~doc)
+
+let metrics_flag =
+  let doc = "Use Metrics; note that it has an impact on performance" in
+  let env = env_var "WITH_METRICS" in
+  Arg.(value & flag & info [ "m"; "with_metrics" ] ~env ~doc)
+
+let log_size =
+  let doc = "The log size of the index." in
+  let env = env_var "LOG_SIZE" in
+  Arg.(value & opt int 500_000 & info [ "log_size" ] ~env ~doc)
+
+let nb_entries =
+  let doc = "The number of bindings." in
+  let env = env_var "NB_ENTRIES" in
+  Arg.(value & opt int 10_000_000 & info [ "nb_entries" ] ~env ~doc)
+
+let list_cmd =
+  let doc = "List all available benchmarks." in
+  (Term.(pure list_benches $ const ()), Term.info "list" ~doc)
+
+let json_flag =
+  let doc = "Output the results as a json object." in
+  let env = env_var "JSON" in
```

Comment:
I think it would be easier to use the json output if it was written in a file. The option `-j` could ask for a filename, for instance?

Reply:
I don't really like the idea of a `--json` flag - which sounds like a format flag -  that also kind of silently affects the output channel.
What would you think of a separate `-o`/`--output` option that lets you specify the output file? We could combine that with the `--json` flag to print the json output in a file.

Reply:
Yeah, that sounds good to me.

Reply:
Done :heavy_check_mark: